modvendor/modvendor.go

```diff
@@ -0,0 +1,223 @@
+	}
+
+	logger.Trace().Msg("moving cloned mod from workdir to clonedir")
+	if err := os.Rename(workdir, clonedir); err != nil {
+		// This may leave intermediary created dirs hanging on vendordir
+		// since we just create all and then delete clone dir on a failure to move.
+		// If we get a lot of errors from os.Rename we may need to handle this
+		// more gracefully, here we assume that os.Rename errors are rare since both
+		// dirs where just created.
+		errs := errors.L()
+		errs.Append(errors.E(err, "moving cloned module"))
+		errs.Append(os.Remove(clonedir))
+		return "", errs.AsError()
+	}
+	return clonedir, nil
+}
+
+// ParseSource parses the given modsource string.
+// The modsource must be a valid Terraform Git/Github source reference as documented in:
+//
+// - https://www.terraform.io/language/modules/sources
+//
+// Source references that are not Git/Github are not supported.
+func ParseSource(modsource string) (Source, error) {
+	ref := ""
+	splitParams := strings.Split(modsource, "?")
+	modsource = splitParams[0]
+
+	if len(splitParams) > 1 {
+		if len(splitParams) != 2 {
+			return Source{}, errors.E(ErrInvalidModSrc, "unexpected extra '?' on source")
+		}
+
+		refParam := splitParams[1]
+		splitRefParam := strings.Split(refParam, "=")
+
+		if len(splitRefParam) != 2 {
+			return Source{}, errors.E(ErrInvalidModSrc, "parsing ref param %q", refParam)
+		}
+
+		if splitRefParam[0] != "ref" {
+			return Source{}, errors.E(ErrInvalidModSrc, "unknown param %q", splitRefParam[0])
+		}
+
+		ref = splitRefParam[1]
+		if ref == "" {
+			return Source{}, errors.E(ErrInvalidModSrc, "ref param %q is empty", refParam)
+		}
+	}
+
+	switch {
+	case strings.HasPrefix(modsource, "github.com"):
+		return Source{
+			URL:  fmt.Sprintf("https://%s.git", modsource),
+			Path: modsource,
+			Ref:  ref,
+		}, nil
+	case strings.HasPrefix(modsource, "git@"):
+		// In git it could be any user@host, but here we are supporting
+		// the specific options allowed by Terraform on module.source:
+		// - https://www.terraform.io/language/modules/sources#github
+		// In this case being Github ssh access.
+		return Source{
+			URL:  modsource,
+			Path: parseGithubAtPath(modsource),
+			Ref:  ref,
+		}, nil
+	case strings.HasPrefix(modsource, "git::"):
+		// Generic git: https://www.terraform.io/language/modules/sources#generic-git-repository
+		path, err := parseGitColonPath(modsource)
+		if err != nil {
+			return Source{}, err
+		}
+		return Source{
+			URL:  strings.TrimPrefix(modsource, "git::"),
+			Path: path,
+			Ref:  ref,
+		}, nil
+
+	default:
+		return Source{}, errors.E(ErrUnsupportedModSrc)
+	}
+}
+
+func parseGithubAtPath(modsource string) string {
+	path := strings.TrimPrefix(modsource, "git@")
+	path = strings.TrimSuffix(path, ".git")
+	// This is GH specific, so we don't need to handle specific ports
+	// and we can assume it is always git@github.com:org/path.git.
+	return strings.Replace(path, ":", "/", 1)
```

Comment:
is this safe? what if user provide a wrong URL like `git@github.com:22/owner/repo.git` ?
maybe we can check here if there's really one instance of `:` in the source and fail if that's not the case?

Reply:
Checking for only one instance of : doesn't solve the issue, when we have a port there will be only one :. It would require a more detailed parsing

Reply:
they are two issues. The first question is not related to the proposed change.

Reply:
I proposed to check for one instance of `:` because the strings.Replace() in this case is intended to replace only the `<host>:<path>` separator, so it should fail if it would replace more stuff.

Reply:
Currently what would happen is that the port number would be part of the path, which is not "unsafe" (it wont break) but also not desireable :laughing:. Will handle removing the port when it is present. Will add the check for how much : we have too...I dont think it is allowed on URLs to have : unless encoded on the path,. but better to ensure on a check indeed.

Reply:
I answered but it didn't show your comments before, sorry :laughing:. Yeah later I got that indeed there is two issues, thanks for pointing out.

Currently on the code I replace only the first ":", initially I thought that would be enough (then the path would have : on the dir names):

```go
strings.Replace(path, ":", "/", 1)
```

Reply:
Will try to improve the parsing in general, thanks for pointing this stuff out :heart: